packages/back-end/src/models/BaseModel.ts

```diff
@@ -84,94 +88,150 @@ export abstract class BaseModel<T extends BaseSchema> {
   /***************
    * Optional methods that can be overridden by subclasses as needed
    ***************/
+  protected async filterByReadPermissions(
+    docs: z.infer<T>[]
+  ): Promise<z.infer<T>[]> {
+    await this.populateForeignRefs(docs);
+
+    const filtered: z.infer<T>[] = [];
+    for (const doc of docs) {
+      if (this.canRead(doc)) {
+        filtered.push(doc);
+      }
+    }
+    return filtered;
+  }
   protected migrate(legacyDoc: unknown): z.infer<T> {
     return legacyDoc as z.infer<T>;
   }
-  protected async customValidation(doc: z.infer<T>) {
+  protected async customValidation(
+    doc: z.infer<T>,
+    writeOptions?: WriteOptions
+  ) {
     // Do nothing by default
   }
-  protected async beforeCreate(doc: z.infer<T>) {
+  protected async beforeCreate(doc: z.infer<T>, writeOptions?: WriteOptions) {
     // Do nothing by default
   }
-  protected async afterCreate(doc: z.infer<T>) {
+  protected async afterCreate(doc: z.infer<T>, writeOptions?: WriteOptions) {
     // Do nothing by default
   }
   protected async beforeUpdate(
     existing: z.infer<T>,
     updates: UpdateProps<z.infer<T>>,
-    newDoc: z.infer<T>
+    newDoc: z.infer<T>,
+    writeOptions?: WriteOptions
   ) {
     // Do nothing by default
   }
   protected async afterUpdate(
     existing: z.infer<T>,
     updates: UpdateProps<z.infer<T>>,
-    newDoc: z.infer<T>
+    newDoc: z.infer<T>,
+    writeOptions?: WriteOptions
   ) {
     // Do nothing by default
   }
-  protected async beforeDelete(doc: z.infer<T>) {
+  protected async beforeDelete(doc: z.infer<T>, writeOptions?: WriteOptions) {
+    // Do nothing by default
+  }
+  protected async afterDelete(doc: z.infer<T>, writeOptions?: WriteOptions) {
     // Do nothing by default
   }
-  protected async afterDelete(doc: z.infer<T>) {
+  protected async afterCreateOrUpdate(
+    doc: z.infer<T>,
+    writeOptions?: WriteOptions
+  ) {
     // Do nothing by default
   }
 
+  protected getForeignKeys(doc: z.infer<T>): ForeignKeys {
+    const keys: ForeignKeys = {};
+
+    // Experiment
+    const experiment = this.detectForeignKey(doc, [
+      "experiment",
+      "experimentId",
+    ]);
+    if (experiment) {
+      keys.experiment = experiment;
+    }
+
+    // Datasource
+    const datasource = this.detectForeignKey(doc, [
+      "datasource",
+      "datasourceId",
+    ]);
+    if (datasource) {
+      keys.datasource = datasource;
+    }
+
+    return keys;
+  }
+
   /***************
    * Built-in public methods
    ***************/
   public getById(id: string) {
+    if (typeof id !== "string") {
+      throw new Error("Invalid id");
+    }
+    if (!id) return Promise.resolve(null);
+
     return this._findOne({ id });
   }
   public getByIds(ids: string[]) {
+    // Make sure ids is an array of strings
+    if (!Array.isArray(ids) || !ids.every((id) => typeof id === "string")) {
```

Comment:
You don't trust TS?

Reply:
A lot of times, we just type the incoming `Request` body without doing proper validation.  These sanity checks are meant to protect against security vulnerabilities related to that.  For example, someone passing in an "id" as `{ "$exists": true }` and our code is assuming it's a string and naively querying Mongo with `{ id: idFromReq }`.